[WIP] Migration of MindMath activity from Vue2 to Vue3

S-kus opened the pull request.

llaske commented on activities/MindMath.activity/js/activity.js:
```diff
@@ -7,14 +7,26 @@ requirejs.config({
+    'sugar-icon': SugarIcon
   },
-  data: function() {
+  data() {
```
Are you sure this syntax should be changed?
I guess `data: function() {` means the same.

llaske commented on activities/MindMath.activity/js/activity.js:
```diff
@@ -1314,3 +1326,6 @@ var app = new Vue({
 });
+
+app.mount('#app');
+console.log(app);
```
Remove log message

llaske commented on activities/MindMath.activity/js/activity.js:
```diff
@@ -7,14 +7,26 @@ requirejs.config({
-var app = new Vue({
-  el: '#app',
+const app = Vue.createApp({
+  // el: '#app',
```
Remove commented line

llaske reviewed (changes requested):
Nice.

llaske commented:
It doesn't work on my side. There is error in the console.

![image](https://user-images.githubusercontent.com/981841/177406302-305ff59d-0982-4909-b7ba-08d7a4de1f52.png)

S-kus commented:
Hey @llaske, this task is still in progress!
I will correct the suggestions with the next commit.

S-kus commented:
## Migration of Pawn.activity from Vue 2 to Vue 3

1. **Error:** App is not starting, and components are not accessible anymore, because of the old syntax of declaring the app and mounting it.
	**Solution:** Declare the app using `const app = Vue.createApp{}`  and mount it using `app.mount('#app');`
2. **Error:** Other components are not mounted correctly. Facing error of `Failed to resolve component`.
    **Solution:** Register the components we are using. For this, declare Sugar components as `const  ComponentName=  {}` by replacing old method of `Vue.component('component-name',  {});` and register them in the main app, so that we can use them globally, as 
    ```
    components: {
	'component-name': ComponentName
	},
    ``` 
  3. **Error:** `Vue.functionName` will give errors. 
    **Solution:** As we decalred our App as `app`, we need to update where we are using `Vue.functionName()` to `app.functionName()`, For exa: `app.directve()`
    4. **Error:** `$children` and `$on` property can no longer be used in the Vue3 app. It's used for `initialization` and `activity.setup()` in our activity.
    **Solution:** We can correct the use of `$children` by using `refs` but for `$on` we need to do some correction. For this:
    In the `SugarLocalization.mounted` method add this:
    
   ```
	    mounted: function () {
	        var vm = this;
	        if (vm.l10n == null) {
	            requirejs(["sugar-web/env", "webL10n"], function (env, webL10n) {
	                env.getEnvironment(function (err, environment) {
	                    vm.l10n = webL10n;
	                    var defaultLanguage = (typeof chrome != 'undefined' && chrome.app && chrome.app.runtime) ? chrome.i18n.getUILanguage() : navigator.language;
	                    var language = environment.user ? environment.user.language : defaultLanguage;
	                    webL10n.language.code = language;
	                    window.addEventListener("localized", function () {
	                        if (!vm.eventReceived) {
	                            vm.code = language;
	                            vm.dictionary = vm.l10n.dictionary;
	                        } else if (webL10n.language.code != language) {
	                            webL10n.language.code = language;
	                        }
	                    });
	                });
	            });
	            //Activity initialization check
	            var activityCheck = function() {
	                var SugarActivity = vm.$root.$refs.SugarActivity;
	                if (SugarActivity.getEnvironment()) {
	                    vm.activityInitialized = true;
	                } else {
	                    window.setTimeout(activityCheck, 100);
	                }
	            }
	            window.setTimeout(activityCheck, 100);
	        }
	    },
  ```
    
and in In `SugarTutorial.mounted` method add the following code:
    
   ```
	    mounted() {
	        let vm = this;
	        var localizeCheck = function() {
	            var SugarL10n = vm.$root.$refs.SugarL10n;
	            if (SugarL10n.activityInitialized) {
	                SugarL10n.localize(vm.l10n)
	            } else {
	                window.setTimeout(localizeCheck, 100);
	            }
	        }
	        window.setTimeout(localizeCheck, 100);
	    },
	    
```

5. **Warnings:**  Faced many warnings related to Vue3 syntax. 
	a. custom `directives`  should be decalred in `directives:` , for this removed `v-visible` directive decalration from `created()` of `SugarToolbar.js` and added like this in same file:
	```
	directives: {
		visible: function  (el,  binding)  {
			// v-visible="condition" (Use this with palettes to avoid errors)
			el.style.visibility = !!binding.value ? 'visible' : 'hidden';
		};
	},
	```
	b. `$listeners` are no more required the Event listeners are now part of `$attrs` so, Removed all usages of `$listeners`.
	c. Warning about missing `template` or `render` function can be solved by creating an empty `render` function like this: ` render() {} `, as we don't need any template for our components.

llaske commented:
@prakashujjwal1010 we're attempting to migrate MindMath to Vue3. Are you able to help us on it?

prakashujjwal1010 commented:
Yeah, sure. Give me some time to get my head around it.

prakashujjwal1010 commented:
Folowing are  the changes required (mindmath specific):
1) Watchers on arrays need to have `deep:true` e,g, In `Game.js`
      ```
      slots: function(newVal) {
            var vm = this;
            //deselecting
            vm.deselect();
      }
      ```
      should be converted to
      ```
      slots: {
            handler(){
              var vm = this;
              //deselecting
              vm.deselect();
            },
            deep: true
          }
      ```
     similar changes are required in `Leaderboard.js` (`playersAll` , `playersPlaying` watchers) and in `activity.js` (`slots` , `compulsoryOps` , `playersPlaying` watchers)
2) `destroyed` lifecycle option should be renamed to `unmounted` in `Result.js`, `Leaderboard.js` and `Game.js`.
3) In `Game.js` 
     - remove `l10n: {
        stringScore: ''
      }` from the data function and also remove or delete the `localize` method.
     - change `v-on:click.native="onSelectNumber(index)"` to `v-on:click="onSelectNumber(index)"`.
4) In `Slots.js` change `<template>` to `<div>` tag.
5) In `index.html` ,
    - close all the Link tags e.g `<link rel="stylesheet" href="css/activity.css" >` to `<link rel="stylesheet" href="css/activity.css" />`
    - change
        ```
             <transition name="fade" id="app-content" mode="out-in">
			      <component
				      .....
			      ></component>
	      </transition>
        ```
       to
       ```
         <div id="app-content">
		      <transition name="fade" mode="out-in">
				      <component
					      .....
				      ></component>
		      </transition>
	      </div>
      ```


6) In `Result.js` and `Leaderboard.js` add new method `generateXOLogoWithColor`:
   ```
        generateXOLogoWithColor: function (strokeColor, fillColor) {
            var xoLogo = '<?xml version="1.0" ?><!DOCTYPE svg  PUBLIC \'-//W3C//DTD SVG 1.1//EN\'  \'http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\' [<!ENTITY stroke_color "#010101"><!ENTITY fill_color "#FFFFFF">]><svg enable-background="new 0 0 55 55" height="55px" version="1.1" viewBox="0 0 55 55" width="55px" x="0px" xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" y="0px"><g display="block" id="stock-xo_1_"><path d="M33.233,35.1l10.102,10.1c0.752,0.75,1.217,1.783,1.217,2.932   c0,2.287-1.855,4.143-4.146,4.143c-1.145,0-2.178-0.463-2.932-1.211L27.372,40.961l-10.1,10.1c-0.75,0.75-1.787,1.211-2.934,1.211   c-2.284,0-4.143-1.854-4.143-4.141c0-1.146,0.465-2.184,1.212-2.934l10.104-10.102L11.409,24.995   c-0.747-0.748-1.212-1.785-1.212-2.93c0-2.289,1.854-4.146,4.146-4.146c1.143,0,2.18,0.465,2.93,1.214l10.099,10.102l10.102-10.103   c0.754-0.749,1.787-1.214,2.934-1.214c2.289,0,4.146,1.856,4.146,4.145c0,1.146-0.467,2.18-1.217,2.932L33.233,35.1z" fill="&fill_color;" stroke="&stroke_color;" stroke-width="3.5"/><circle cx="27.371" cy="10.849" fill="&fill_color;" r="8.122" stroke="&stroke_color;" stroke-width="3.5"/></g></svg>';
            var coloredLogo = xoLogo;
            coloredLogo = coloredLogo.replace("#010101", strokeColor)
            coloredLogo = coloredLogo.replace("#FFFFFF", fillColor)
            return "data:image/svg+xml;base64," + btoa(coloredLogo);
          }
   ```
    and you can remove `generateXOLogoWithColor` function definition from `utils/utils.js`.
7) In `activity.js` in `handleRestartButton` method, change `if (vm.SugarPresence.isHost) {` to `if (vm.SugarPresence && vm.SugarPresence.isHost) {`

### Finals step
Now after making all the above changes, you will find that the activity is still not running.
So ,In `activity.js` in `mounted()` try commenting out `vm.SugarPresence = vm.$refs.SugarPresence;` line.
```
//vm.SugarPresence = vm.$refs.SugarPresence;
```
Now try to run the activity.. voila it works (but now you can't use sugar presence T_T, obviously ) . 
so we can see that the `SugarPresence` component is the one breaking the activity. I would advice @S-kus , start testing `SugarPresence` component. I will try to test it at my end also.

llaske commented:
Good job @prakashujjwal1010.
BTW I still have an error message.

![image](https://user-images.githubusercontent.com/981841/179356656-6aee45b0-93cf-483a-a9f2-b2e016c54d40.png)

It seems to come from the following line in `activity.js` :

`
vm.strokeColor = vm.currentenv.user.colorvalue.stroke;`

@S-kus do you have the same issue?

prakashujjwal1010 commented on activities/MindMath.activity/index.html:
```diff
@@ -88,41 +88,41 @@
-	</transition>
-
+		<div id="app-content">
+			<transition name="fade" id="app-content" mode="out-in">
```
remove `id` from `<transition>` i.e. change it to `<transition name="fade" mode="out-in">`

S-kus commented:
> remove id from transition 

Thanks, @prakashujjwal1010, Corrected this! Now it's working fine on my side.
@llaske You can review this again!

llaske commented:
@rockharshitmaurya yes, sure.
